### DISCUSSION TITLE
Why: Package versions are represented internally with 4 digits instead of 3

### DISCUSSION THREAD

**Discussion by mxr576** [composer/composer#11456] [Q&A / Support] [answered]
Why Package versions represented internally with 4 digit instead of 3 in Composer? E.g,
* 3.2.0 is 3.2.0.0 
* 1.0.0-RC3 is 1.0.0.0-RC

(For example, when I call `\Composer\Package\PackageInterface::getVersion()`.)

There must be a historical reason behind it :) 

Is there a way to get the 3 digit representation that we can see "everywhere"?

**Comment by stof** [comment, marked as answer]
Composer does not restrict its versions to what semver support. And IIRC, historically, some packages where using 4 segments in their versions.

Given that PHP's `version_compare` does not properly handle version strings with different number of segments, Composer normalizes all versions to the maximum number of segments it supports.
You can use `PackageInterface::getPrettyVersion()` to get the non-normalized version (but there is no guarantee that it will have 3 segments. It depends how you named your tags).

**Reply by mxr576** [reply]
:man_facepalming: I must have tried that before, but somehow I did not see that work... now it works if I have a _package_.

However, on constraints, it does seem to work, I still get the 4 digits version. I am unsure if there is a solution for that too.

https://github.com/mxr576/ddqg-composer-audit/blob/2cc4cd27a64fb48cd6f80b92d09647ae98c0f05a/src/Domain/SecurityAdvisory/SecurityAdvisoryFinderFromProblematicPackageProvider.php#L41C84-L49

**Reply by stof** [reply]
Well, the `->normalize()` call is precisely about taking the pretty version as input and returning the normalized version. So you actually have the pretty version in your code. That's `$packageConstraintMap[$package_name]->getLowerBound()->getVersion()`

**Reply by mxr576** [reply]
>  the ->normalize() call is precisely about taking the pretty version as input and returning the normalized version. 

Yes, `normalize()` returns the normalized version, it is a good point. I will probably remove it because now it also does not make sense to me that it is called.

> That's $packageConstraintMap[$package_name]->getLowerBound()->getVersion()

That also returns the normalized version. I have double-checked it.
![image](https://github.com/composer/composer/assets/1755573/3283002a-7203-4358-a335-025c4309f5c8)
![image](https://github.com/composer/composer/assets/1755573/90b5198a-d4ab-459a-a8ee-aa919307a150)
